### ISSUE TITLE
Add BOS token to calibration dataset in LLMs

### ISSUE THREAD

**Issue by Giuseppe5** [Xilinx/brevitas#1229]
**Is your feature request related to a problem? Please describe.**
Through a fortuitous debug, we realized that adding BOS token to calibration dataset could help, at least in terms of perplexity, in data-dependent PTQ algorithms.
Pending a verification on the effect on few-shot accuracy, we might decide to add this as pre-processing step.

This would break reproducibility of past results.

**Describe the solution you'd like**
Option 1:
We have a breaking change, assuming this is a unequivocal improvement

Option 2:
We have yet another flag, default set to preserve retro-compatibility 

Option 3:
Same as above, but the default corresponds to the new behaviour, and a specific flag is needed to reproduce past results

Personally, I could go for Option 2 (since we haven't had a release in a while, it could be worth having a snapshot of the current behaviour), maybe with a warning that in the next release we will move to option 3 (assuming the change is always an improvement).

**Comment by nickfraser** [comment]
Note, PR #1230 plans to eventually address this issue.

**Comment by nickfraser** [comment]
Suggestion / proposal:

Perhaps we should consider modifying this:

>```
>    batch = data['input_ids'][:, (i * seqlen):((i + 1) * seqlen)]
>```

to break sentences at specific tokens (i.e., EOS, '.', '\n'), this would make it so that the prepending of BOS tokens into subsequent strings make sense. Of course, we would still need to crop the subsequences to match the target `seqlen`. It would add some extra annoyance, but at least it might make the whole pipeline make more sense.

**Comment by nickfraser** [comment]
> Consequently, when partioning according to the seqlen of the model, this token will only be present in the first position of the first batch (i.e. testloader[0]["input_ids][0, 0]). Note that a similar preprocessing is used in [SpinQuant](https://github.com/facebookresearch/SpinQuant/tree/main) and [GPTQ](https://github.com/IST-DASLab/gptq).

Interestingly, the cropping for calibration with these tools selects from [random substrings](https://github.com/IST-DASLab/gptq/blob/669e51587e93a1cc04015119e26b1b00e48e77c9/datautils.py#L20-L30) within each sequence.

**Comment by pablomlago** [comment]
To validate the effect on perplexity when doing preprocessing the calibration/validation datasets using the methodology in [Nanotron](https://github.com/huggingface/nanotron/blob/main/src/nanotron/data/processing.py), we conducted a GPTQ experiment with varying calibration datasets: Legacy Wikitext2 (pre-processing in `dev`), Wikitext2, Pile and C4, while evaluating the perplexity on Wikitext2, either using the pre-processing in dev (Legacy in Table), or the pre-processing in Nanotron.

The results are shown in the following Table:

| Calibration | Legacy Float PPL | Float PPL | Legacy Quant PPL | Quant PPL |
|-------------|----------|-----------|-----------|-----------|
| Legacy Wikitext2     | 8.75    |       | 13.8125    |       |
| Wikitext2 | 8.75    | 18.00     | 13.8125    |  33.000 |
| C4       | 8.75    |      |     |       |
| Pile        | 8.75    |      |     |       |

Note that the Float PPL, corresponding to the Nanotron pre-processing, is considerably higher than the perplexity obtained when evaluating in `dev`. However, the relative increases in perplexity due to quantization are somewhat comparable, i.e. ~1.57x (Legacy Float PPL -> Legacy Quant PPL) and ~1.83x (Float PPL -> Quant PPL), for the row `Wikitext2`. However, we can see that the results match for Wikitext2 and Legacy Wikitext2, when evaluating using the same methodology.

Based on this results, it might be desirable to modify the pre-processing of our calibration datasets, to match the methods used when pre-training the model, while keeping our evaluation as it is, for direct comparison with other quantization works.

**Comment by Giuseppe5** [comment]
Fixed with #1240

**Comment by i-colbert** [comment]
I know this issue is closed, but we may need to push this a little further.

#1240 adds BOS pre-processing to the training/calibration dataset by separately tokenizing the data as independent sentences (thus adding BOS to each sentence), then concatenating, and finally randomly splitting the data to create `nsamples` sequences of length `seqlen`. However, this is not also done for the validation set. I would argue that if we pre-process the training/calibration dataset this way, then we would also need to pre-process the validation set by at least separately tokenizing the data as independent sentences (thus adding BOS to each sentence).

Furthermore, BOS typically stands for "beginning of sequence" not "sentence," and so it may make sense to instead pre-process by adding BOS to the beginning of every random sequence and not mixing it in with every sentence. Indeed, some [recent work](https://arxiv.org/pdf/2504.02732) has concluded this from some interesting experimentation on attention sinks:
> If the model is trained with ⟨bos⟩ always appearing at the first token, removing ⟨bos⟩ during inference removes the attention sink, i.e. the model relies on the ⟨bos⟩ token to form the sink. Instead, if there is no ⟨bos⟩ during training, the sink forms at the first token regardless, but is slightly weaker.

This is what was originally done for Brevitas, but the "pre-processing" was only done in the evaluation function. 

From this, we have 2 different types of BOS pre-processing in Brevitas:
1. Sentence-level BOS pre-processing (seems like nanotron does this)
2. Sequence-level BOS pre-processing (seems like Llama3 does this)

Right now, dev does a mix-and-match of both: sentence-level BOS pre-processing is added only to the training/calibration dataset if specified (note it is on by default) and never to the validation dataset, and sequence-level BOS pre-processing is added in the evaluation function (note this does not touch the training/calibration dataset). So, the training dataset uses sentence-level BOS pre-processing, but the validation dataset uses sequence-level BOS pre-processing.

My suggestion is to support both of these BOS pre-processing schemes as mutually exclusive options for both the training/calibration dataset and the validation dataset and avoid mixing and matching. For example, I would disable the sentence-level BOS pre-processing for Llama3 but enable the sequence-level BOS pre-processing. Here is some data to support such a decision, which is also supported by the theoretical conclusions in the linked paper on attention sinks.

|       | float (train PPL) | float (test PPL) | quant-base (train PPL) | quant-base (test PPL) | quant-base (0-shot acc.) | quant-gptq (train PPL) | quant-gptq (test PPL) | quant-gptq (0-shot acc.) |
|-------|-------|-------|------------|------------|------------|------------|------------|------------|
| 1B    | 25.8  | 8.9   | 216.0      | 18.0       | 49.1       | 35.2       | 10.8       | 55.6       |
| 3B    | 77.0  | 7.2   | 128.0      | 10.4       | 60.8       | 105.0      | 8.4        | 64.9       |


This is 4-bit weight-only asymmetric integer quantization with floating-point scaling factors without any PTQ techniques. Note that the training/calibration dataset is sentence-level BOS pre-processing while the test set is not.

Now, #1258 is aiming to remove the sequence-level BOS pre-processing from the evaluation function. The sequence-level BOS pre-processing is likely better to be done in the data pre-processing functions and not the evaluation function, which is how we got into this line of questioning.

**Comment by Giuseppe5** [comment]
> (2) sequence-level BOS pre-processing is added in the evaluation function (note this does not touch the training dataset)
This is currently being removed.

The current implementation aims at replicating closely what other SOTA quantization algorithms are using to produce their result, as to have numbers that are more or less comparable (even though they cannot be overlapped 100%).

Once the last BOS PR is merged, and considering your results above, my feeling is that we are not here to optimize for perplexity or 0-shot evaluation per se. 
We need a reliable and logical approach to compute these metrics, and I believe the future version of dev after #1258 will get us there.

We always re-compute our own baselines anyway, and the important part is that the relative comparisons hold true.

As you pointed out, there is no single solution that seems to work best across the board for both ppl and 0-shot, and having more options to play with is not always better, especially within Brevitas where our search space is already massive on every possible direction.


What you propose would also directly conflict with the future dev, where we wouldn't have anymore a "base" validation set with no preprocessing applied to it, which again is what most of the SOTA algorithms do. How would that fit into this dichotomy between sequence BOS and sentence BOS?

As much as I find this interesting, it is not a quantization-related issue, and in an ideal world I would have only one solution, which is what everyone else is using (currently we already have 2 in dev, and your proposal doesn't strictly match with either one so maybe we need to jump to four?).


Having said that, if you feel like working on this, I would need a few more experiments with different models and sizes to fully evaluate its impact, and I have not the bandwidth at the moment for that.

**Comment by i-colbert** [comment]
It may be too early to say that this issue isn't related to quantization techniques that rely on calibration data. The preliminary results provided earlier suggest BOS pre-processing non-trivially impacts GPTQ, for example.

There doesn't seem to be a standard for BOS pre-processing. Although [nanotron seems to use sentence-level BOS pre-processing](https://github.com/huggingface/nanotron/blob/37cc8b0cd4795d6b458f609f9671c5c465e09333/src/nanotron/data/processing.py#L22), a cursory dive into LightEval shows that [special tokens are added by default for LightEval](https://github.com/huggingface/lighteval/blob/69931dfd2435c85e1efdf89e00b2b9f0d06e5d11/src/lighteval/models/transformers/transformers_model.py#L142), and some light debuging shows that they are BOS tokens added to the beginning of each sequence (not within each sentence).

Here is some more data for Llama3.2 1B with the different forms of BOS pre-processing and their impact on the training and testing datasets for the BF16 reference models.

| preproc | train | test |
|---------|------|------|
| none | 9.19 | 8.94 |
| sequence | 9.19 | 8.94 |
| sentence |25.75 |23.88 |

I have opened a PR to implement the suggested evaluation support for sentence- or sequence-level pre-processing, with a default to doing nothing: #1272

**Comment by fxmarty-amd** [comment]
@i-colbert @Giuseppe5 In a specific case, I happen to see detrimental results from applying GPTQ when:

1. Using bos_token in the train set
2. Not using bos_token in evaluation

**Comment by i-colbert** [comment]
> [@i-colbert](https://github.com/i-colbert) [@Giuseppe5](https://github.com/Giuseppe5) In a specific case, I happen to see detrimental results from applying GPTQ when:
> 
> 1. Using bos_token in the train set
> 2. Not using bos_token in evaluation

Ya, this is expected. GPTQ is data-driven, and the bos_token impacts the activation distribution, likely via attention sinks (see references from this thread). You want your test set to be IID with your calibration set. If you pre-process one and not the other, this is not the case. An analog in image classification would be normalizing your training set but not your test set.

**Comment by fxmarty-amd** [comment]
Thanks @i-colbert, makes sense.
